### DISCUSSION TITLE
Can't find resource decoratorcontroller  metacontroller v2.0.4 in kubernetes v1.22.0

### DISCUSSION THREAD

**Discussion by nevzatgul** [metacontroller/metacontroller#357] [Q&A]
I want to create service per pod in kubernetes v1.22.  My cluster have multiple shard.
I have created  a cluster with  GoogleCloudPlatform  metacontroller 0.4.0 and  kubernetes v1.21 before. 
After upgrading to kubernetes 1.22, I had to change the versions from v1beta1 to v1.
In order to update version to v1, I had to update metacontroller too. So I decided to update GoogleCloudPlatform  metacontroller 0.4.0 to metacontroller v2.0.4.

When I want to create decoratorcontroller with the yaml files for version v2.0.4  which are published inside examples on your github repository , I get the following error for shard 0.

kubernetes@km1:~$ kubectl describe decoratorcontrollers.metacontroller.k8s.io  shard-0-pod-name-label
..
..
..
Events:
 Type     Reason       Age                   From            Message
  ----     ------       ----                  ----            -------
  Warning  CreateError  16h (x20 over 16h)    metacontroller  Cannot create new controller: can't find resource "pods" in apiVersion "v1"
  Warning  CreateError  27m (x20 over 66m)    metacontroller  Cannot create new controller: can't find resource "pods" in apiVersion "v1"
  Warning  CreateError  62s (x15 over 2m24s)  metacontroller  Cannot create new controller: can't find resource "pods" in apiVersion "v1"

--

kubernetes@km1:~$ kubectl describe decoratorcontrollers.metacontroller.k8s.io  shard-0-service-per-pod
..
..
..
Events:
  Type     Reason       Age                   From            Message
  ----     ------       ----                  ----            -------
  Warning  CreateError  16h (x20 over 16h)    metacontroller  Cannot create new controller: can't find resource "statefulsets" in apiVersion "apps/v1"
  Warning  CreateError  28m (x20 over 66m)    metacontroller  Cannot create new controller: can't find resource "statefulsets" in apiVersion "apps/v1"
  Warning  CreateError  69s (x15 over 2m31s)  metacontroller  Cannot create new controller: can't find resource "statefulsets" in apiVersion "apps/v1"

--

kubernetes@km1:~$ kubectl logs -n metacontroller metacontroller-0

{"level":"error","ts":1630050326.0129108,"logger":"controller-runtime.manager.controller.decorator-metacontroller","msg":"Reconciler error","name":"shard-0-pod-name-label","namespace":"","error":"can't find resource \"pods\" in apiVersion \"v1\"","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.5/pkg/internal/controller/controller.go:253\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.5/pkg/internal/controller/controller.go:214"}
{"level":"Level(-4)","ts":1630050326.0129387,"logger":"decorator","msg":"Sync DecoratorController","name":"shard-0-service-per-pod"}
{"level":"error","ts":1630050326.013003,"logger":"controller-runtime.manager.controller.decorator-metacontroller","msg":"Reconciler error","name":"shard-0-service-per-pod","namespace":"","error":"can't find resource \"statefulsets\" in apiVersion \"apps/v1\"","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.5/pkg/internal/controller/controller.go:253\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.5/pkg/internal/controller/controller.go:214"}

I don't know  why doesn't  it find any statefulsets.

**Comment by grzesuav** [comment]
hi @nevzatgul , this message is often happening at the start onf metacontroller and/or cache sync, it is described there - https://metacontroller.github.io/metacontroller/guide/troubleshooting.html#common-log-messages.

Except log message, does the example did not work ?

**Reply by TobiasGoerke** [reply]
I'd understand this behavior when dealing with CRDs which, as described in the documentation you've linked, were recently installed. 
However, in @nevzatgul's case, metacontroller doesn't find core resources like pods and statefulsets. I'm seeing the same log messages at startup and don't think they should occur..

Btw, in v3.0.1 the log message has changed to the following. Apparently, metacontroller assumes resources to be CRDs. 

> {"level":"error","ts":1650872552.3326426,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"my-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}

**Reply by grzesuav** [reply]
and it keeps like that ? Or they self-resolve ? How look like Decorator/Composite controller definition in that case ?

**Reply by TobiasGoerke** [reply]
The log messages stop appearing after some time. 

I've compiled a short example you can run and see for yourself. This snippet  installs Kubeflow, which in return deploys metacontroller. 
It's the quickest I could come up with, sorry if the example is somewhat overblown.

```
git clone https://github.com/kubeflow/manifests.git -b v1.5-branch
cd manifests/example/
kustomize edit set image docker.io/metacontrollerio/metacontroller:v3.0.1
while ! kustomize build | kubectl apply -f -; do echo "Retrying to apply resources"; sleep 10; done
sleep 10
kubectl -n kubeflow wait --for=condition=ready po metacontroller-0
kubectl -n kubeflow logs metacontroller-0 -f 
```

Edit: this is the CompositeController the output is referring to: https://github.com/kubeflow/manifests/blob/v1.5.0/apps/pipeline/upstream/base/installs/multi-user/pipelines-profile-controller/composite-controller.yaml

The output then looks like this:
`{"level":"info","ts":1651127200.8949916,"msg":"Configuration information","discovery-interval":3600,"cache-flush-interval":1800,"metrics-address":":9999","client-go-qps":5,"client-go-burst":10,"workers":5,"events-qps":0.0033333333333333335,"events-burst":25,"pprofAddr":"0","leader-election":false,"leader-election-resource-lock":"leases","leader-election-namespace":"","leader-election-id":"metacontroller","version":"v3.0.1"}
{"level":"info","ts":1651127200.9003828,"msg":"Communication with K8s API server successful"}
{"level":"info","ts":1651127202.1016986,"msg":"Waited for 1.1921176s due to client-side throttling, not priority and fairness, request: GET:https://10.43.0.1:443/apis/serving.kubeflow.org/v1beta1?timeout=32s\n"}
{"level":"info","ts":1651127212.1022172,"msg":"Waited for 11.191172s due to client-side throttling, not priority and fairness, request: GET:https://10.43.0.1:443/apis/admissionregistration.k8s.io/v1beta1?timeout=32s\n"}
{"level":"info","ts":1651127222.3015082,"msg":"Waited for 5.9966759s due to client-side throttling, not priority and fairness, request: GET:https://10.43.0.1:443/apis/autoscaling/v2beta1?timeout=32s\n"}
{"level":"info","ts":1651127232.3017595,"msg":"Waited for 15.9968688s due to client-side throttling, not priority and fairness, request: GET:https://10.43.0.1:443/apis/apiextensions.k8s.io/v1beta1?timeout=32s\n"}
{"level":"info","ts":1651127232.9050007,"logger":"controller-runtime.metrics","msg":"Metrics server is starting to listen","addr":":9999"}
{"level":"info","ts":1651127232.9067974,"msg":"Starting server","path":"/metrics","kind":"metrics","addr":"[::]:9999"}
{"level":"info","ts":1651127232.9074311,"logger":"controller.composite-metacontroller","msg":"Starting EventSource","source":"kind source: *v1alpha1.CompositeController"}
{"level":"info","ts":1651127232.9074936,"logger":"controller.composite-metacontroller","msg":"Starting Controller"}
{"level":"info","ts":1651127232.9075818,"logger":"controller.decorator-metacontroller","msg":"Starting EventSource","source":"kind source: *v1alpha1.DecoratorController"}
{"level":"info","ts":1651127232.9076707,"logger":"controller.decorator-metacontroller","msg":"Starting Controller"}
{"level":"info","ts":1651127233.0080936,"logger":"controller.composite-metacontroller","msg":"Starting workers","worker count":1}
{"level":"info","ts":1651127233.0081334,"logger":"controller.decorator-metacontroller","msg":"Starting workers","worker count":1}
{"level":"info","ts":1651127233.0082107,"logger":"composite","msg":"Sync CompositeController","name":"kubeflow-pipelines-profile-controller"}
{"level":"error","ts":1651127233.608949,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"kubeflow-pipelines-profile-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1651127233.614409,"logger":"composite","msg":"Sync CompositeController","name":"kubeflow-pipelines-profile-controller"}
{"level":"error","ts":1651127233.6145859,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"kubeflow-pipelines-profile-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1651127233.6250286,"logger":"composite","msg":"Sync CompositeController","name":"kubeflow-pipelines-profile-controller"}
{"level":"error","ts":1651127233.6251924,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"kubeflow-pipelines-profile-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1651127233.645736,"logger":"composite","msg":"Sync CompositeController","name":"kubeflow-pipelines-profile-controller"}
{"level":"error","ts":1651127233.6458845,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"kubeflow-pipelines-profile-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1651127233.686227,"logger":"composite","msg":"Sync CompositeController","name":"kubeflow-pipelines-profile-controller"}
{"level":"error","ts":1651127233.6863391,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"kubeflow-pipelines-profile-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1651127233.7669442,"logger":"composite","msg":"Sync CompositeController","name":"kubeflow-pipelines-profile-controller"}
{"level":"error","ts":1651127233.7670789,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"kubeflow-pipelines-profile-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1651127233.9279687,"logger":"composite","msg":"Sync CompositeController","name":"kubeflow-pipelines-profile-controller"}
{"level":"error","ts":1651127233.9281132,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"kubeflow-pipelines-profile-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1651127234.2489817,"logger":"composite","msg":"Sync CompositeController","name":"kubeflow-pipelines-profile-controller"}
{"level":"error","ts":1651127234.2490807,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"kubeflow-pipelines-profile-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1651127234.8898969,"logger":"composite","msg":"Sync CompositeController","name":"kubeflow-pipelines-profile-controller"}
{"level":"error","ts":1651127234.890036,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"kubeflow-pipelines-profile-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1651127236.1708915,"logger":"composite","msg":"Sync CompositeController","name":"kubeflow-pipelines-profile-controller"}
{"level":"error","ts":1651127236.1710234,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"kubeflow-pipelines-profile-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1651127238.731245,"logger":"composite","msg":"Sync CompositeController","name":"kubeflow-pipelines-profile-controller"}
{"level":"error","ts":1651127238.731477,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"kubeflow-pipelines-profile-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1651127242.306924,"msg":"Waited for 9.3968013s due to client-side throttling, not priority and fairness, request: GET:https://10.43.0.1:443/apis/apiregistration.k8s.io/v1beta1?timeout=32s\n"}
{"level":"info","ts":1651127243.8520339,"logger":"composite","msg":"Sync CompositeController","name":"kubeflow-pipelines-profile-controller"}
{"level":"error","ts":1651127243.852185,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"kubeflow-pipelines-profile-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1651127252.3074908,"msg":"Waited for 3.9980184s due to client-side throttling, not priority and fairness, request: GET:https://10.43.0.1:443/apis/authentication.k8s.io/v1?timeout=32s\n"}
{"level":"info","ts":1651127254.0929284,"logger":"composite","msg":"Sync CompositeController","name":"kubeflow-pipelines-profile-controller"}
{"level":"error","ts":1651127254.0930843,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"kubeflow-pipelines-profile-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1651127262.5069947,"msg":"Waited for 14.1973791s due to client-side throttling, not priority and fairness, request: GET:https://10.43.0.1:443/apis/install.istio.io/v1alpha1?timeout=32s\n"}
{"level":"Level(-4)","ts":1651127264.90836,"msg":"Failed to fetch all resources, continuing with partial discovery info","failures":"unable to retrieve the complete list of server APIs: metrics.k8s.io/v1beta1: the server is currently unable to handle the request"}
{"level":"info","ts":1651127274.5733268,"logger":"composite","msg":"Sync CompositeController","name":"kubeflow-pipelines-profile-controller"}
{"level":"error","ts":1651127274.5734715,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"kubeflow-pipelines-profile-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1651127315.5341344,"logger":"composite","msg":"Sync CompositeController","name":"kubeflow-pipelines-profile-controller"}
{"level":"error","ts":1651127315.53429,"logger":"controller.composite-metacontroller","msg":"Reconciler error","name":"kubeflow-pipelines-profile-controller","namespace":"","error":"CustomResourceDefinition.apiextensions.k8s.io \"namespaces.\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}`

**Reply by TobiasGoerke** [reply]
Update: I was wrong - the newer versions of Metacontroller not only throw errors but also won't reconcile correctly (as can be seen in example above - without changing metacontroller's version there are two deployments in the namespace kubeflow-user-example-com)

**Reply by grzesuav** [reply]
@TobiasGoerke on which k8s version you run this  ?

**Reply by TobiasGoerke** [reply]
> @TobiasGoerke on which k8s version you run this ?

I've tested the above example using Rancher Desktop (version 1.2.1) and a fresh Kubernetes install of v1.21.11. 

We've also observed this issue on an OpenShift Cluster using version v1.20.0
